scipy/stats/tests/test_morestats.py

```diff
@@ -2975,90 +2981,96 @@ def test_directional_stats_correctness(self):
+        full_array = xp.asarray(np.tile(data, (2, 2, 2, 1)))
+        expected = xp.asarray([[[1., 0., 0.],
+                                [1., 0., 0.]],
+                               [[1., 0., 0.],
+                                [1., 0., 0.]]],
+                              dtype=xp.float64)
         dirstats = stats.directional_stats(full_array, axis=2)
-        assert_allclose(expected, dirstats.mean_direction)
+        xp_assert_close(dirstats.mean_direction, expected)
 
-    def test_directional_stats_list_ndarray_input(self):
+    def test_directional_stats_list_ndarray_input(self, xp):
         # test that list and numpy array inputs yield same results
         data = [[0.8660254, 0.5, 0.], [0.8660254, -0.5, 0]]
-        data_array = np.asarray(data)
-        res = stats.directional_stats(data)
-        ref = stats.directional_stats(data_array)
-        assert_allclose(res.mean_direction, ref.mean_direction)
-        assert_allclose(res.mean_resultant_length,
-                        res.mean_resultant_length)
-
-    def test_directional_stats_1d_error(self):
+        data_array = xp.asarray(data, dtype=xp.float64)
+        ref = stats.directional_stats(data)
+        res = stats.directional_stats(data_array)
+        xp_assert_close(res.mean_direction,
+                        xp.asarray(ref.mean_direction))
+        xp_assert_close(res.mean_resultant_length,
+                        xp.asarray(res.mean_resultant_length))
+
+    def test_directional_stats_1d_error(self, xp):
         # test that one-dimensional data raises ValueError
-        data = np.ones((5, ))
+        data = xp.ones((5, ))
         message = (r"samples must at least be two-dimensional. "
                    r"Instead samples has shape: (5,)")
         with pytest.raises(ValueError, match=re.escape(message)):
             stats.directional_stats(data)
 
-    def test_directional_stats_normalize(self):
+    def test_directional_stats_normalize(self, xp):
         # test that directional stats calculations yield same results
         # for unnormalized input with normalize=True and normalized
         # input with normalize=False
-        data = np.array([[0.8660254, 0.5, 0.],
-                         [1.7320508, -1., 0.]])
+        data = xp.asarray([[0.8660254, 0.5, 0.],
+                           [1.7320508, -1., 0.]])
         res = stats.directional_stats(data, normalize=True)
-        normalized_data = data / np.linalg.norm(data, axis=-1,
-                                                keepdims=True)
-        ref = stats.directional_stats(normalized_data,
-                                      normalize=False)
-        assert_allclose(res.mean_direction, ref.mean_direction)
-        assert_allclose(res.mean_resultant_length,
-                        ref.mean_resultant_length)
+        # get `xp.linalg.vector_norm` for `xp_vector_norm`
+        xp_test = array_namespace(data)
+        data_norm = xp_vector_norm(data, axis=-1, keepdims=True, xp=xp_test)
```

Comment:
Since `xp_vector_norm` is not tested separately from `directional_stats` right now, I'd feel more comfortable if this test were to continue computing the `normalized_data` with NumPy. (Could just compute the whole reference value with NumPy. If that means specifying `dtype` to make it easier, that's fine.)

Reply:
addressed